File: src/cpp/ee/services/internal/LazyBannerAd.cpp
//
//  LazyAdView.cpp
//  ee-x-d542b565
//
//  Created by eps on 1/7/21.
//

#include "ee/services/internal/LazyBannerAd.hpp"

#include <ee/core/NoAwait.hpp>
#include <ee/core/ObserverHandle.hpp>
#include <ee/core/Task.hpp>

namespace ee {
namespace services {
using Self = LazyBannerAd;

Self::LazyBannerAd() {
    visible_ = false;
    anchor_ = std::pair(0, 0);
    position_ = std::pair(0, 0);
    size_ = std::nullopt;
    handle_ = std::make_unique<ObserverHandle>();
}

Self::~LazyBannerAd() = default;

void Self::setAd(const std::shared_ptr<IBannerAd>& ad) {
    handle_->clear();
    if (ad == nullptr) {
        return;
    }
    (*handle_) //
        .bind(*ad)
        .addObserver({
            .onLoaded =
                [this] {
                    if (visible_) {
                        ad_->setVisible(false);
                        ad_->setVisible(true);
                    }

                    // Propagation.
                    dispatchEvent([](auto&& observer) {
                        if (observer.onLoaded) {
                            observer.onLoaded();
                        }
                    });
                },
            .onClicked =
                [this] {
                    // Propagation.
                    dispatchEvent([](auto&& observer) {
                        if (observer.onClicked) {
                            observer.onClicked();
                        }
                    });
                },
        });
    ad_ = ad;
    ad->setVisible(visible_);
    ad->setAnchor(anchor_.first, anchor_.second);
    ad->setPosition(position_.first, position_.second);
    if (size_.has_value()) {
        ad->setSize(size_->first, size_->second);
    }
    noAwait(ad_->load());
}

void Self::destroy() {
    if (ad_ != nullptr) {
        ad_->destroy();
    }
    handle_->clear();
}

bool Self::isLoaded() const {
    if (ad_ == nullptr) {
        return false;
    }
    return ad_->isLoaded();
}

Task<bool> Self::load() {
    if (ad_ == nullptr) {
        co_return false;
    }
    co_return co_await ad_->load();
}

std::pair<float, float> Self::getAnchor() const {
    return anchor_;
}

void Self::setAnchor(float x, float y) {
    anchor_ = std::pair(x, y);
    if (ad_ != nullptr) {
        ad_->setAnchor(x, y);
    }
}

std::pair<float, float> Self::getPosition() const {
    return position_;
}

void Self::setPosition(float x, float y) {
    position_ = std::pair(x, y);
    if (ad_ != nullptr) {
        ad_->setPosition(x, y);
    }
}

std::pair<float, float> Self::getSize() const {
    if (size_.has_value()) {
        return size_.value();
    }
    if (ad_ != nullptr) {
        return ad_->getSize();
    }
    return std::pair(0, 0);
}

void Self::setSize(float width, float height) {
    size_ = std::pair(width, height);
    if (ad_ != nullptr) {
        ad_->setSize(width, height);
    }
}

bool Self::isVisible() const {
    return visible_;
}

void Self::setVisible(bool visible) {
    visible_ = visible;
    if (ad_ != nullptr) {
        ad_->setVisible(visible);
    }
}
} // namespace services
} // namespace ee

File: src/ts/src/services/internal/LazyBannerAd.ts
import {
    AdObserver,
    IBannerAd,
} from "../../ads";
import {
    ObserverHandle,
    ObserverManager,
    Utils,
} from "../../core";

export class LazyBannerAd extends ObserverManager<AdObserver> implements IBannerAd {
    private _ad?: IBannerAd;
    private _visible: boolean;
    private _anchor: [number, number];
    private _position: [number, number];
    private _size?: [number, number];
    private readonly _handle: ObserverHandle;

    public constructor() {
        super();
        this._visible = false;
        this._anchor = [0, 0];
        this._position = [0, 0];
        this._size = undefined;
        this._handle = new ObserverHandle();
    }

    public get ad(): IBannerAd | undefined {
        return this._ad;
    }

    public set ad(value: IBannerAd | undefined) {
        this._handle.clear();
        if (value === undefined) {
            return;
        }
        this._handle.bind(value)
            .addObserver({
                onLoaded: () => {
                    if (this._visible) {
                        value.isVisible = false;
                        value.isVisible = true;
                    }
                    this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded());
                },

                onLoadResult: result => this.dispatchEvent(observer =>
                    observer.onLoadResult && observer.onLoadResult(result)),

                onClicked: () => this.dispatchEvent(observer =>
                    observer.onClicked && observer.onClicked()),
            });
        this._ad = value;
        this._ad.isVisible = this._visible;
        this._ad.anchor = this._anchor;
        this._ad.position = this._position;
        if (this._size !== undefined) {
            this._ad.size = this._size;
        }
        Utils.noAwait(async () => {
            await value.load();
        });
    }

    public destroy(): void {
        this._ad?.destroy();
        this._handle.clear();
    }

    public get isLoaded(): boolean {
        return this._ad?.isLoaded ?? false;
    }

    public async load(): Promise<boolean> {
        if (this._ad == undefined) {
            return false;
        }
        return await this._ad.load();
    }

    public get anchor(): [number, number] {
        return this._anchor;
    }

    public set anchor(value: [number, number]) {
        this._anchor = value;
        if (this._ad !== undefined) {
            this._ad.anchor = value;
        }
    }

    public get position(): [number, number] {
        return this._position;
    }

    public set position(value: [number, number]) {
        this._position = value;
        if (this._ad !== undefined) {
            this._ad.position = value;
        }
    }

    public get size(): [number, number] {
        if (this._size !== undefined) {
            return this._size;
        }
        return this._ad?.size ?? [0, 0];
    }

    public set size(value: [number, number]) {
        this._size = value;
        if (this._ad !== undefined) {
            this._ad.size = value;
        }
    }

    public get isVisible(): boolean {
        return this._visible;
    }

    public set isVisible(value: boolean) {
        this._visible = value;
        if (this._ad !== undefined) {
            this._ad.isVisible = value;
        }
    }
}